Make the base fee configurable in ChannelConfig

TheBlueMatt opened the pull request:
Currently the base fee we apply is always the expected cost to
claim an HTLC on-chain in case of closure. This results in
significantly higher than market rate fees [1], and doesn't really
match the actual forwarding trust model anyway - as long as
channel counterparties are honest, our HTLCs shouldn't end up
on-chain no matter what the HTLC sender/recipient do.

While some users may wish to use a feerate that implies they will
not lose funds even if they go to chain (assuming no flood-and-loot
style attacks), they should do so by calculating fees themselves;
since they're already charging well above market-rate,
over-estimating some won't have a large impact.

This commit adds a configuration knob to set the base fee
explicitly, defaulting to 1 sat, which appears to be market-rate
today.

[1] Note that due to an msat-vs-sat bug we currently actually
    charge 1000x *less* than the calculated cost.

TheBlueMatt commented:
0.1 since none of our immediate users need payment forwarding, but without this we will spuriously fail payments.

TheBlueMatt commented:
Oh whoops, this needs #970.

TheBlueMatt commented:
Moving this back to 0.0.99 as I have another PR that should be based on it which needs to go in 0.0.99.

ariard commented on lightning/src/util/config.rs:
```diff
@@ -224,6 +236,23 @@ pub struct UserConfig {
 	pub peer_channel_config_limits: ChannelHandshakeLimits,
 	/// Channel config which affects behavior during channel lifetime.
 	pub channel_options: ChannelConfig,
+	/// If a channel is not set to be publicly announced, we reject HTLCs which were set to be
```
nit: maybe we should dissociate HTLC "rejection" at `update_add_htlc` processing from "forward failure", what do you think about "fail forward" ?

TheBlueMatt replied:
You mean in the docs? I feel like "reject" communicates whats going on, especially to a user who doesn't know about internal distinctions of HTLC rejections. "we fail HTLCs which were set" is slightly less clear to me, but also has roughly the same meaning in non-technical english.

ariard replied:
Well, i think it's yet another anticipation of introducing custom forward policy (`PendingHTLCsRelayable`), which would a new source of rejection cases. We'll cleanup at that time i guess.

ariard commented on lightning/src/util/config.rs:
```diff
@@ -224,6 +236,23 @@ pub struct UserConfig {
+	/// ensure you are not exposed to any forwarding risk.
+	///
+	/// Note that because you cannot change a channel's announced state after creation, there is no
+	/// way to disable forwarding on public channels retroactively. Thus, in order to change a node
```
I think #680 by allowing to load custom routing policy would make it easy to disallow forwarding in-flight on a named public channels by dropping the HTLC. Yep iptables for LN routing nodes!

TheBlueMatt replied:
Right, if users really want to do so, I guess we should let them, but they really shouldn't, ideally, want to do so - if you aren't "really" routing payments, your channel shouldn't be public. I wonder if we should only apply #680 to routing out over private channels? Anyway, we can debate it there.

ariard reviewed (commented):
Some attacks discussions though overall sounds good to me

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1521,15 +1525,23 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+						// Note that the behavior here should be identical to the above block - we
+						// should NOT reveal the existence or non-existence of a private channel if
+						// we don't allow forwards outbound over them.
+						break Some(("Don't have available channel for forwarding as requested.", 0x4000 | 10, None));
```
Right, for something with `accept_forwards_to_priv_channels` set we do have such a timing attack. At some point we need to move more things behind the forwarding delay (including claims, this, etc), its just low on the priority list :/.

ariard replied:
Tracked in #680, yeah agree low-priority for now...

TheBlueMatt commented on lightning/src/util/config.rs:
```diff
@@ -146,6 +149,13 @@ pub struct ChannelConfig {
+	/// update messages sent to notify all nodes of our updated relay fee.
+	///
+	/// Default value: 1000.
+	pub fee_base_msat: u32,
```
> Maybe add a note that default setting is matching market-rate as of June 2021

done.

> Also, we could add a different logic where the fee_base_msat increase in function of the channel load, thus making the last remaining HTLC slots more expensive to harden the liquidity cost of jamming the channel and divert the jamming flow to cheaper nodes ? Though I concede it's hard to evaluate utility of such mitigation without precise a jamming attacker, especially laying out attacker economic rationality of targeting this node.

Yea, there's a few things I'd love to do with the feebase, but it doesn't belong in `ChannelManager`, IMO. I've been idly thinking of a separate crate that "manages" some stuff like this for routing nodes - auto-rebalances channels by intercepting payment events before handing them to the user, manages feerates based on various criteria, etc. Its a conversation for later, though.

> Though i agree with you on not caring about fee_base_msat for flood-and-lood attacks, the cost risk of going onchain with any HTLC should be accounted as part of channel mitigation with some kind of compensation

Yea, but ultimately I think this is a question for the spec and pre-paying a fee followed by a fee rebate when its claimed, or something like that.

ariard replied:
> I've been idly thinking of a separate crate that "manages" some stuff like this for routing nodes - auto-rebalances channels by intercepting payment events before handing them to the user, manages feerates based on various criteria, etc. Its a conversation for later, though.

Yeah we're aligned on this, better to get channel liquidity management in another crate!

> Yea, but ultimately I think this is a question for the spec and pre-paying a fee followed by a fee rebate when its claimed, or something like that.

Yeap, same here beyond the scope of our implementation. Just to be sure we agree on what this new `fee_base_msat` is not covering.

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -3477,24 +3476,8 @@ impl<Signer: Sign> Channel<Signer> {
-		res += fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::Normal) as u64 * SPENDING_INPUT_FOR_A_OUTPUT_WEIGHT / 1000;
-
-		res as u32
+	pub fn get_holder_fee_base_msat(&self) -> u32 {
```
Hmm, I do want to make sure its communicated that its *our* forwarding fee, not the fee our counterparty charges to forward to us. How about `get_outbound_forwarding_fee_base_msat`?

ariard replied:
Yeah sounds a good name to me `get_outbound_forwarding_fee_base_msat`!

TheBlueMatt commented:
Rebased on latest git now that #970 is merged.

valentinewallace commented on lightning/src/util/config.rs:
```diff
@@ -237,6 +240,23 @@ pub struct UserConfig {
 	pub peer_channel_config_limits: ChannelHandshakeLimits,
 	/// Channel config which affects behavior during channel lifetime.
 	pub channel_options: ChannelConfig,
+	/// If a channel is not set to be publicly announced, we reject HTLCs which were set to be
+	/// forwarded over the channel if this is set to false. This prevents use from taking on
+	/// HTLC-forwarding risk when we intend to run as a node which is not online reliably.
```
Maybe clearer phrasing: If this is set to false, we'll refuse to forward HTLCs over any private channels we create.

valentinewallace commented on lightning/src/util/config.rs:
```diff
@@ -237,6 +240,23 @@ pub struct UserConfig {
 	/// Channel config which affects behavior during channel lifetime.
 	pub channel_options: ChannelConfig,
+	/// If a channel is not set to be publicly announced, we reject HTLCs which were set to be
+	/// forwarded over the channel if this is set to false. This prevents use from taking on
```
nit: s/use/us

valentinewallace commented on lightning/src/util/config.rs:
```diff
@@ -245,6 +265,7 @@ impl Default for UserConfig {
 			own_channel_config: ChannelHandshakeConfig::default(),
 			peer_channel_config_limits: ChannelHandshakeLimits::default(),
 			channel_options: ChannelConfig::default(),
+			accept_forwards_to_priv_channels: false,
```
Would this be better suited for the ChannelConfig on channel creation, to make it configurable per-channel?

TheBlueMatt replied:
Hmmm, is there a specific use-case you have in mind? I feel like this is kinda just a "do you want forwarding" knob, the only reason its not exactly that is we should generally never allow the disabling of forwarding out over public channels, cause its realllllyyyy anti-social.

valentinewallace replied:
No particular use-case in mind. Seems a little more "future-proof" but eh. Resolving

TheBlueMatt commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -1341,22 +1341,27 @@ pub fn create_node_cfgs<'a>(node_count: usize, chanmon_cfgs: &'a Vec<TestChanMon
+	default_config.channel_options.cltv_expiry_delta = 6*6;
+	default_config.channel_options.announced_channel = true;
+	default_config.peer_channel_config_limits.force_announced_channel_preference = false;
+	default_config.own_channel_config.our_htlc_minimum_msat = 1000; // sanitization being done by the sender, to exerce receiver logic we need to lift of limit
```
I believe it was meant as `exercise`, but I have no idea what it means overall lol. Maybe @ariard remembers?

ariard replied:
IIRC, it's related to `test_override_0msat_htlc_minimum`, where we want to pass 0-msat value as a config and verify that our implementation-sanitization logic at channel opening bumps correctly to 1. See commit 9e03d2bc. It's a bit gross, rational is for covering spec requirement in BOLT2 

```
A receiving node:

    receiving an amount_msat equal to 0, OR less than its own htlc_minimum_msat:
        SHOULD fail the channel.
``` 

Yes `exerce` is to be understood as `exercise`, pardon my french :p

TheBlueMatt replied:
Thanks! Updated the comment to describe it.

TheBlueMatt commented on lightning/src/ln/onion_route_tests.rs:
```diff
@@ -240,17 +241,64 @@ impl Writeable for BogusOnionHopData {
-	use ln::channelmanager::CLTV_FAR_FAR_AWAY;
-	use bitcoin::secp256k1;
+fn test_fee_failures() {
+	// When we check for amount_below_minimum below, we want to test that we're using the *right*
```
Its copied from the below test, but, indeed, doesn't make sense here.

TheBlueMatt commented on lightning/src/util/config.rs:
```diff
@@ -146,6 +146,17 @@ pub struct ChannelConfig {
 	pub fee_proportional_millionths: u32,
+	/// Amount (in milli-satoshi) the channel will charge for each transferred HTLC, in excess of
+	/// fee_proportional_millionths.
+	/// This may be allowed to change at runtime in a later update, however doing so must result in
```
Restart doesn't update this config, either. They're set in stone unless you open new channels, for now. We really need to make them able to change, but we haven't done so yet.

valentinewallace commented on lightning/src/ln/onion_route_tests.rs:
```diff
@@ -240,17 +241,54 @@ impl Writeable for BogusOnionHopData {
-	const PERM: u16 = 0x4000;
-	const NODE: u16 = 0x2000;
-	const UPDATE: u16 = 0x1000;
+	// In an earlier version, we spuriously failed to forward payments if the expected feerate
+	// changed between the channel open and the payment.
```
Is this a reference to the `msat-vs-sat bug` mentioned in the commit msg? Maybe missing something, but where can I find more information about the `msat-vs-sat bug`?

TheBlueMatt replied:
No, literally current code is "we decide what we think the fee should be when we receive a payment to forward, and dont bother to update our announcements".....

TheBlueMatt replied:
Oops, sorry, an earlier version of this patchset had separated out the bugfixes for the two bugs, but now they're in the same, I've updated the commit message to be more clear.

valentinewallace commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -7895,6 +7914,151 @@ fn test_announce_disable_channels() {
+
+	create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 1_000_000, 500_000_000, InitFeatures::known(), InitFeatures::known());
+
+	nodes[1].node.create_channel(nodes[2].node.get_our_node_id(), 1_000_000, 500_000_000, 42, None).unwrap();
+	let open_channel = get_event_msg!(nodes[1], MessageSendEvent::SendOpenChannel, nodes[2].node.get_our_node_id());
+	nodes[2].node.handle_open_channel(&nodes[1].node.get_our_node_id(), InitFeatures::known(), &open_channel);
+	let accept_channel = get_event_msg!(nodes[2], MessageSendEvent::SendAcceptChannel, nodes[1].node.get_our_node_id());
+	nodes[1].node.handle_accept_channel(&nodes[2].node.get_our_node_id(), InitFeatures::known(), &accept_channel);
+
+	let (temporary_channel_id, tx, _) = create_funding_transaction(&nodes[1], 1_000_000, 42);
+	nodes[1].node.funding_transaction_generated(&temporary_channel_id, tx.clone()).unwrap();
+	nodes[2].node.handle_funding_created(&nodes[1].node.get_our_node_id(), &get_event_msg!(nodes[1], MessageSendEvent::SendFundingCreated, nodes[2].node.get_our_node_id()));
+	check_added_monitors!(nodes[2], 1);
+
+	nodes[1].node.handle_funding_signed(&nodes[2].node.get_our_node_id(), &get_event_msg!(nodes[2], MessageSendEvent::SendFundingSigned, nodes[1].node.get_our_node_id()));
+	check_added_monitors!(nodes[1], 1);
+
+	let conf_height = core::cmp::max(nodes[1].best_block_info().1 + 1, nodes[2].best_block_info().1 + 1);
+	confirm_transaction_at(&nodes[1], &tx, conf_height);
+	connect_blocks(&nodes[1], CHAN_CONFIRM_DEPTH - 1);
+	confirm_transaction_at(&nodes[2], &tx, conf_height);
+	connect_blocks(&nodes[2], CHAN_CONFIRM_DEPTH - 1);
+	nodes[1].node.handle_funding_locked(&nodes[2].node.get_our_node_id(), &get_event_msg!(nodes[2], MessageSendEvent::SendFundingLocked, nodes[1].node.get_our_node_id()));
+	let funding_locked = get_event_msg!(nodes[1], MessageSendEvent::SendFundingLocked, nodes[2].node.get_our_node_id());
+	nodes[2].node.handle_funding_locked(&nodes[1].node.get_our_node_id(), &funding_locked);
+
```
Confused why `create_chan_between_nodes` or similar couldn't have been used here? A comment for this may be useful

TheBlueMatt replied:
It currently expects `announcement_signatures`. I could maybe have split it up more, but it would have required a number of new functions, which didn't really seem worth it?

valentinewallace commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -7895,6 +7914,151 @@ fn test_announce_disable_channels() {
+		channel_monitors.insert(monitor_a.get_funding_txo().0, &mut monitor_a);
+		channel_monitors.insert(monitor_b.get_funding_txo().0, &mut monitor_b);
+		<(BlockHash, ChannelManager<EnforcingSigner, &test_utils::TestChainMonitor, &test_utils::TestBroadcaster, &test_utils::TestKeysInterface, &test_utils::TestFeeEstimator, &test_utils::TestLogger>)>::read(&mut nodes_1_read, ChannelManagerReadArgs {
+			default_config: no_announce_cfg,
```
Just to be clear, users couldn't do exactly what this test is doing, to change their config on restart?

TheBlueMatt replied:
Nope, users can change the non-channel parts of the config at restart (but not at runtime...).

valentinewallace commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -7895,6 +7914,151 @@ fn test_announce_disable_channels() {
+	expect_payment_failed!(nodes[0], our_payment_hash, false);
+	expect_payment_failure_chan_update!(nodes[0], nodes[2].node.list_channels()[0].short_channel_id.unwrap(), true);
+
+	nodes[0].node.peer_disconnected(&nodes[1].node.get_our_node_id(), false);
```
Maybe a comment here: "We now disconnect and restart RL in order to change the channel config to allow forwards over private channels"

valentinewallace commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -7895,6 +7914,151 @@ fn test_announce_disable_channels() {
+	let node_cfgs = create_node_cfgs(3, &chanmon_cfgs);
+	let mut no_announce_cfg = test_default_channel_config();
+	no_announce_cfg.channel_options.announced_channel = false;
+	no_announce_cfg.accept_forwards_to_priv_channels = false;
```
Guess we shouldn't need this because it's `false` by default?

TheBlueMatt replied:
Sure, I just like making the test robust against a change to default values in the future.

ariard commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1108,6 +1108,10 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+	///
+	/// Note that we do not check if you are currently connected to the given peer. If no
+	/// connection is available, the outbound `open_channel` message may fail to send, resulting in
+	/// the channel eventually being silently forgotten.
```
What do you mean by "being silently forgotten", can't remember this case is covered by `peer_disconnected` ?

TheBlueMatt replied:
It is, it will be silently forgotten when we disconnect or restart.

TheBlueMatt commented:
Rebased and added a commit (that only makes sense if we merge this for 0.0.99) to use the same deserialization version check as here in the fields from #984.

valentinewallace commented on lightning/src/util/config.rs:
```diff
@@ -146,6 +149,17 @@ pub struct ChannelConfig {
+	/// update messages sent to notify all nodes of our updated relay fee.
+	///
+	/// The default value of a single satoshi roughly matches the market rate on many routing nodes
+	/// as of July 2021. Ajusting it upwards or downwards may change whether nodes route through
```
nit: Adjusting

TheBlueMatt commented:
Squashed with only the typo fix changed:
```
$ git diff-tree -U1 cb68789 09d51ea5
diff --git a/lightning/src/util/config.rs b/lightning/src/util/config.rs
index d738eefd..10d98deb 100644
--- a/lightning/src/util/config.rs
+++ b/lightning/src/util/config.rs
@@ -157,3 +157,3 @@ pub struct ChannelConfig {
 	/// The default value of a single satoshi roughly matches the market rate on many routing nodes
-	/// as of July 2021. Ajusting it upwards or downwards may change whether nodes route through
+	/// as of July 2021. Adjusting it upwards or downwards may change whether nodes route through
 	/// this node.
$
```

naumenkogs commented:
Concept ACK defaulting to 1 sat :)

ariard commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -7914,6 +7915,156 @@ fn test_announce_disable_channels() {
+
+	assert!(nodes[0].node.list_usable_channels()[0].is_public);
+	assert_eq!(nodes[1].node.list_usable_channels().len(), 2);
+	assert!(!nodes[2].node.list_usable_channels()[0].is_public);
```
Hmmmm, i was suggesting to add the following diff but I do have non-deterministic failure, i think it's inherited from `list_usable_channels()`

```
diff --git a/lightning/src/ln/functional_tests.rs b/lightning/src/ln/functional_tests.rs
index 29071ab0..11154f5c 100644
--- a/lightning/src/ln/functional_tests.rs
+++ b/lightning/src/ln/functional_tests.rs
@@ -7960,6 +7960,8 @@ fn test_priv_forwarding_rejection() {
 
        assert!(nodes[0].node.list_usable_channels()[0].is_public);
        assert_eq!(nodes[1].node.list_usable_channels().len(), 2);
+       assert!(nodes[1].node.list_usable_channels()[0].is_public);
+       assert!(!nodes[1].node.list_usable_channels()[1].is_public);
        assert!(!nodes[2].node.list_usable_channels()[0].is_public);
 
        // We should always be able to forward through nodes[1] as long as its out through a public
```

TheBlueMatt replied:
The ordering of the channels is non-deterministic - its based on the random seed used for the hashmap, even, so this will fail randomly.

ariard commented on lightning/src/ln/channel.rs:
```diff
@@ -4844,20 +4844,25 @@ impl<'a, Signer: Sign, K: Deref> ReadableArgs<&'a K> for Channel<Signer>
+			minimum_depth = Some(Readable::read(reader)?);
+		} else {
+			// Read the 4 bytes of backwards-compatibility data.
+			let _dummy: u32 = Readable::read(reader)?;
```
Could be nice to have the corresponding comment in the `Writer`, at `minimum_depth`/`counterparty_selected_channel_reserve_satoshis`, "write the field for backwards-compatibility data" ?

ariard reviewed (commented):
ACK 09d51ea, i did verify new functionality correctness and exercise their test coverage.

Few comments/remarks but no blockers for this PR, can be addressed in follow-ups.

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -4460,7 +4460,7 @@ fn is_unsupported_shutdown_script(their_features: &InitFeatures, script: &Script
 }
 
-const SERIALIZATION_VERSION: u8 = 1;
+const SERIALIZATION_VERSION: u8 = 2;
```
Yea, we're kinda still figuring it out and doing a case-by-case "what do you *want* the old clients to do". Def could add to CONTRIBUTING that we have to meticulously document any backwards compat concerns in the CHANGELOG :)

TheBlueMatt commented on lightning/src/util/config.rs:
```diff
@@ -197,6 +208,7 @@ impl Default for ChannelConfig {
 	fn default() -> Self {
 		ChannelConfig {
 			fee_proportional_millionths: 0,
+			fee_base_msat: 1000,
```
Hmm, well that's mostly cause `test_priv_forwarding_rejection` isn't intended to test the default fee value :). The onion_route_tests and others will fail if we don't properly check the forwarding fee, though.

TheBlueMatt commented on lightning/src/util/config.rs:
```diff
@@ -146,6 +146,17 @@ pub struct ChannelConfig {
+	/// this node.
+	///
+	/// Default value: 1000.
+	pub fee_base_msat: u32,
```
I just mentioned its the outbound leg in the docs, adding it to the variable name seemed like a useless distinction.

TheBlueMatt commented:
Squashed a few trivial changes to address all review feedback, will take after CI. Full diff from Val's review:
```
$ git diff-tree -U1  cb68789  f557df99
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index 93197d49..87f4e6ea 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -3427,3 +3427,3 @@ impl<Signer: Sign> Channel<Signer> {
 	pub fn get_fee_proportional_millionths(&self) -> u32 {
-		self.config.fee_proportional_millionths
+		self.config.forwarding_fee_proportional_millionths
 	}
@@ -3501,3 +3501,3 @@ impl<Signer: Sign> Channel<Signer> {
 	pub fn get_outbound_forwarding_fee_base_msat(&self) -> u32 {
-		self.config.fee_base_msat
+		self.config.forwarding_fee_base_msat
 	}
@@ -4490,3 +4490,3 @@ impl<Signer: Sign> Writeable for Channel<Signer> {
 		// deserializers, but we will read the version in the TLV at the end instead.
-		self.config.fee_proportional_millionths.write(writer)?;
+		self.config.forwarding_fee_proportional_millionths.write(writer)?;
 		self.config.cltv_expiry_delta.write(writer)?;
@@ -4652,3 +4652,6 @@ impl<Signer: Sign> Writeable for Channel<Signer> {
 		self.counterparty_max_htlc_value_in_flight_msat.write(writer)?;
+
+		// Note that this field is ignored by 0.0.99+ as the TLV Optional variant is used instead.
 		self.counterparty_selected_channel_reserve_satoshis.unwrap_or(0).write(writer)?;
+
 		self.counterparty_htlc_minimum_msat.write(writer)?;
@@ -4656,2 +4659,4 @@ impl<Signer: Sign> Writeable for Channel<Signer> {
 		self.counterparty_max_accepted_htlcs.write(writer)?;
+
+		// Note that this field is ignored by 0.0.99+ as the TLV Optional variant is used instead.
 		self.minimum_depth.unwrap_or(0).write(writer)?;
@@ -4709,3 +4714,3 @@ impl<'a, Signer: Sign, K: Deref> ReadableArgs<&'a K> for Channel<Signer>
 			// Read the old serialization of the ChannelConfig from version 0.0.98.
-			config.as_mut().unwrap().fee_proportional_millionths = Readable::read(reader)?;
+			config.as_mut().unwrap().forwarding_fee_proportional_millionths = Readable::read(reader)?;
 			config.as_mut().unwrap().cltv_expiry_delta = Readable::read(reader)?;
diff --git a/lightning/src/ln/functional_tests.rs b/lightning/src/ln/functional_tests.rs
index 29071ab0..986ff9d8 100644
--- a/lightning/src/ln/functional_tests.rs
+++ b/lightning/src/ln/functional_tests.rs
@@ -1900,3 +1900,3 @@ fn test_channel_reserve_holding_cell_htlcs() {
 	let mut config = test_default_channel_config();
-	config.channel_options.fee_base_msat = 239;
+	config.channel_options.forwarding_fee_base_msat = 239;
 	let node_chanmgrs = create_node_chanmgrs(3, &node_cfgs, &[Some(config.clone()), Some(config.clone()), Some(config.clone())]);
@@ -5353,3 +5353,3 @@ fn test_duplicate_payment_hash_one_failure_one_success() {
 	let mut config = test_default_channel_config();
-	config.channel_options.fee_base_msat = 196;
+	config.channel_options.forwarding_fee_base_msat = 196;
 	let node_chanmgrs = create_node_chanmgrs(4, &node_cfgs,
@@ -5545,3 +5545,3 @@ fn do_test_fail_backwards_unrevoked_remote_announce(deliver_last_raa: bool, anno
 	let mut config = test_default_channel_config();
-	config.channel_options.fee_base_msat = 196;
+	config.channel_options.forwarding_fee_base_msat = 196;
 	let node_chanmgrs = create_node_chanmgrs(6, &node_cfgs,
@@ -6400,3 +6400,3 @@ fn test_fail_holding_cell_htlc_upon_free_multihop() {
 	let mut config = test_default_channel_config();
-	config.channel_options.fee_base_msat = 196;
+	config.channel_options.forwarding_fee_base_msat = 196;
 	let node_chanmgrs = create_node_chanmgrs(3, &node_cfgs, &[Some(config.clone()), Some(config.clone()), Some(config.clone())]);
@@ -7968,2 +7968,4 @@ fn test_priv_forwarding_rejection() {
 
+	// ... however, if we send to nodes[2], we will have to pass the private channel from nodes[1]
+	// to nodes[2], which should be rejected:
 	let (our_payment_preimage, our_payment_hash, our_payment_secret) = get_payment_preimage_hash!(nodes[2]);
diff --git a/lightning/src/ln/onion_route_tests.rs b/lightning/src/ln/onion_route_tests.rs
index e325ffce..62086430 100644
--- a/lightning/src/ln/onion_route_tests.rs
+++ b/lightning/src/ln/onion_route_tests.rs
@@ -257,3 +257,3 @@ fn test_fee_failures() {
 	let mut config = test_default_channel_config();
-	config.channel_options.fee_base_msat = 196;
+	config.channel_options.forwarding_fee_base_msat = 196;
 
@@ -308,3 +308,3 @@ fn test_onion_failure() {
 	let mut config = test_default_channel_config();
-	config.channel_options.fee_base_msat = 196;
+	config.channel_options.forwarding_fee_base_msat = 196;
 
diff --git a/lightning/src/util/config.rs b/lightning/src/util/config.rs
index d738eefd..f8535000 100644
--- a/lightning/src/util/config.rs
+++ b/lightning/src/util/config.rs
@@ -145,3 +145,4 @@ impl Default for ChannelHandshakeLimits {
 pub struct ChannelConfig {
-	/// Amount (in millionths of a satoshi) the channel will charge per transferred satoshi.
+	/// Amount (in millionths of a satoshi) charged per satoshi for payments forwarded outbound
+	/// over the channel.
 	/// This may be allowed to change at runtime in a later update, however doing so must result in
@@ -150,5 +151,5 @@ pub struct ChannelConfig {
 	/// Default value: 0.
-	pub fee_proportional_millionths: u32,
-	/// Amount (in milli-satoshi) the channel will charge for each transferred HTLC, in excess of
-	/// fee_proportional_millionths.
+	pub forwarding_fee_proportional_millionths: u32,
+	/// Amount (in milli-satoshi) charged for payments forwarded outbound over the channel, in
+	/// excess of [`forwarding_fee_proportional_millionths`].
 	/// This may be allowed to change at runtime in a later update, however doing so must result in
@@ -157,3 +158,3 @@ pub struct ChannelConfig {
 	/// The default value of a single satoshi roughly matches the market rate on many routing nodes
-	/// as of July 2021. Ajusting it upwards or downwards may change whether nodes route through
+	/// as of July 2021. Adjusting it upwards or downwards may change whether nodes route through
 	/// this node.
@@ -161,3 +162,3 @@ pub struct ChannelConfig {
 	/// Default value: 1000.
-	pub fee_base_msat: u32,
+	pub forwarding_fee_base_msat: u32,
 	/// The difference in the CLTV value between incoming HTLCs and an outbound HTLC forwarded over
@@ -212,4 +213,4 @@ impl Default for ChannelConfig {
 		ChannelConfig {
-			fee_proportional_millionths: 0,
-			fee_base_msat: 1000,
+			forwarding_fee_proportional_millionths: 0,
+			forwarding_fee_base_msat: 1000,
 			cltv_expiry_delta: 6 * 12, // 6 blocks/hour * 12 hours
@@ -222,3 +223,3 @@ impl Default for ChannelConfig {
 impl_writeable_tlv_based!(ChannelConfig, {
-	(0, fee_proportional_millionths, required),
+	(0, forwarding_fee_proportional_millionths, required),
 	(2, cltv_expiry_delta, required),
@@ -226,3 +227,3 @@ impl_writeable_tlv_based!(ChannelConfig, {
 	(6, commit_upfront_shutdown_pubkey, required),
-	(8, fee_base_msat, required),
+	(8, forwarding_fee_base_msat, required),
 });
$
```

TheBlueMatt commented:
Turns out upstream changes made the new tests fail, applied this diff on top of a clean rebase.
```
iff --git a/lightning/src/ln/functional_test_utils.rs b/lightning/src/ln/functional_test_utils.rs
index 8bb49ef9..fe13c64f 100644
--- a/lightning/src/ln/functional_test_utils.rs
+++ b/lightning/src/ln/functional_test_utils.rs
@@ -379,6 +379,31 @@ macro_rules! get_event_msg {
        }
 }
 
+/// Get two specific event messages from the pending events queue.
+#[macro_export]
+macro_rules! get_two_event_msgs {
+       ($node: expr, $event_type_a: path, $event_type_b: path, $node_id: expr) => {
+               {
+                       let events = $node.node.get_and_clear_pending_msg_events();
+                       assert_eq!(events.len(), 2);
+                       (match events[0] {
+                               $event_type_a { ref node_id, ref msg } => {
+                                       assert_eq!(*node_id, $node_id);
+                                       (*msg).clone()
+                               },
+                               _ => panic!("Unexpected event"),
+                       },
+                       match events[1] {
+                               $event_type_b { ref node_id, ref msg } => {
+                                       assert_eq!(*node_id, $node_id);
+                                       (*msg).clone()
+                               },
+                               _ => panic!("Unexpected event"),
+                       })
+               }
+       }
+}
+
 /// Get a specific event from the pending events queue.
 #[macro_export]
 macro_rules! get_event {
diff --git a/lightning/src/ln/functional_tests.rs b/lightning/src/ln/functional_tests.rs
index 8e5e5839..feb8bd7e 100644
--- a/lightning/src/ln/functional_tests.rs
+++ b/lightning/src/ln/functional_tests.rs
@@ -7955,9 +7955,11 @@ fn test_priv_forwarding_rejection() {
        connect_blocks(&nodes[1], CHAN_CONFIRM_DEPTH - 1);
        confirm_transaction_at(&nodes[2], &tx, conf_height);
        connect_blocks(&nodes[2], CHAN_CONFIRM_DEPTH - 1);
+       let as_funding_locked = get_event_msg!(nodes[1], MessageSendEvent::SendFundingLocked, nodes[2].node.get_our_node_id());
        nodes[1].node.handle_funding_locked(&nodes[2].node.get_our_node_id(), &get_event_msg!(nodes[2], MessageSendEvent::SendFundingLocked, nodes[1].node.get_our_node_id()));
-       let funding_locked = get_event_msg!(nodes[1], MessageSendEvent::SendFundingLocked, nodes[2].node.get_our_node_id());
-       nodes[2].node.handle_funding_locked(&nodes[1].node.get_our_node_id(), &funding_locked);
+       get_event_msg!(nodes[1], MessageSendEvent::SendChannelUpdate, nodes[2].node.get_our_node_id());
+       nodes[2].node.handle_funding_locked(&nodes[1].node.get_our_node_id(), &as_funding_locked);
+       get_event_msg!(nodes[2], MessageSendEvent::SendChannelUpdate, nodes[1].node.get_our_node_id());
 
        assert!(nodes[0].node.list_usable_channels()[0].is_public);
        assert_eq!(nodes[1].node.list_usable_channels().len(), 2);
@@ -8054,8 +8056,10 @@ fn test_priv_forwarding_rejection() {
 
        nodes[0].node.peer_connected(&nodes[1].node.get_our_node_id(), &msgs::Init { features: InitFeatures::known() });
        nodes[1].node.peer_connected(&nodes[0].node.get_our_node_id(), &msgs::Init { features: InitFeatures::empty() });
-       nodes[1].node.handle_channel_reestablish(&nodes[0].node.get_our_node_id(), &get_event_msg!(nodes[0], MessageSendEvent::SendChannelReestablish, nodes[1].node.get_our_node_id()));
-       nodes[0].node.handle_channel_reestablish(&nodes[1].node.get_our_node_id(), &get_event_msg!(nodes[1], MessageSendEvent::SendChannelReestablish, nodes[0].node.get_our_node_id()));
+       nodes[1].node.handle_channel_reestablish(&nodes[0].node.get_our_node_id(),
+               &get_two_event_msgs!(nodes[0], MessageSendEvent::SendChannelReestablish, MessageSendEvent::SendChannelUpdate, nodes[1].node.get_our_node_id()).0);
+       nodes[0].node.handle_channel_reestablish(&nodes[1].node.get_our_node_id(),
+               &get_two_event_msgs!(nodes[1], MessageSendEvent::SendChannelReestablish, MessageSendEvent::SendChannelUpdate, nodes[0].node.get_our_node_id()).0);
 
        nodes[1].node.peer_connected(&nodes[2].node.get_our_node_id(), &msgs::Init { features: InitFeatures::known() });
        nodes[2].node.peer_connected(&nodes[1].node.get_our_node_id(), &msgs::Init { features: InitFeatures::empty() });
```
